microscopium/metrics.py

```diff
@@ -154,3 +154,99 @@ def gene_distance_score(X, collection, metric='euclidean'):
     all_intragene_data = distance[all_intragene_index]
     all_intergene_data = distance[all_intergene_index]
     return all_intragene_data, all_intergene_data
+
+
+def _partition_range(values1, values2, n):
+    """
+    Builds a partition of bins over the entire range of values1 and values 2.
+
+    Parameters
+    ----------
+    values1, values2 : arrays
+        arrays to be concatenated
+    n : int
+        number of bins
+
+    Returns
+    -------
+    partition : array
+        A 1D array of bin edges, of length n+1
+    """
+
+    eps = 1e-30
+    d_max = max(np.max(values1), np.max(values2)) + eps
+    d_min = min(np.min(values1), np.min(values2))
+    partition = np.linspace(d_min, d_max, n) #or n, check this
+
+    return partition
+
+
+def _empirical_distribution(values, bins):
+    """
+    Returns an EDF of an input array 'values' over a given array of bin edges
+    Note: this is a pdf, not a cdf
+
+    Parameters
+    ----------
+    values : array
+    bins : bins
+
+
+    Returns
+    -------
+    edf : array
+        A probability distribution of the frequencies of values which appear in the respective bins
+    """
+
+    ind = np.digitize(values, bins)
+
+    #Note: np.digitize bin index starts from index 1
+    #erray returns number of times each data point occurs
+    edf=np.bincount(ind, minlength = len(bins) + 1)
+
+    #normalize
+    edf = edf / np.sum(edf)
+
+    return edf
+
+
+def bhattacharyya_distance(values0, values1, n):
+    """Returns the Bhattacharyya coefficient of 2 input arrays
+    BC of 2 probability distributions, f(x) and g(x), is given by the sum of Sqrt(f(x_i)g(x_i)), over all indices i
```

Comment:
And this line is *definitely* too long. =)

Note that many documentation generators support latex, in between dollar signs, so just write the definition in latex. You can also reference as follows:

```
This is a totally ludicrous statement, but it's backed by evidence [1]_.

[...]

References
----------
.. [1] Jack Sparrow et al, 1842. Study demonstrating ludicrous things.
```